### PR TITLE
Add telemetry metrics to cover the SDK operations

### PR DESCRIPTION
**Pull request by muhammad-othman**
## Description

Added the following metrics to the SDK:
* client.call.duration
* client.uptime
* client.call.attempts
* client.call.errors
* client.call.attempt_duration
* client.call.resolve_endpoint_duration
* client.call.serialization_duration
* client.call.deserialization_duration
* client.call.auth.signing_duration
* client.call.auth.resolve_identity_duration
* client.http.bytes_sent
* client.http.bytes_received

Switched `TelemetryProvider` properties sets to private to not allow subclasses to alter them and making the register methods the only mutable mechanism, and updated `DefaultTelemetryProvider` acordingly.

## Motivation and Context
DOTNET-7518

## Testing
Created a hacked OTel provider, ran same requests and checked the output in the console and visualized these traces in observability tools, here are some examples of the tracked metrics from different services and actions:

<br/>
<details>
  <summary><b>Console examples:</b></summary>

```
Metric Name: client.uptime, Unit: s, Meter: AWSSDK.S3
(2024-07-09T20:51:56.0155210Z, 2024-07-09T20:52:15.4079103Z] rpc.service: S3 rpc.system: aws-api Histogram
Value: Sum: 16.7116956 Count: 1 Min: 16.7116956 Max: 16.7116956
(10,25]:1


Metric Name: client.call.duration, Unit: s, Meter: AWSSDK.S3
(2024-07-09T20:51:56.0631967Z, 2024-07-09T20:52:15.4079126Z] rpc.method: ListBuckets rpc.service: S3 rpc.system: aws-api Histogram
Value: Sum: 0.7526558 Count: 1 Min: 0.7526558 Max: 0.7526558
(0,5]:1

(2024-07-09T20:51:56.0631967Z, 2024-07-09T20:52:15.4079126Z] rpc.method: PutBucket rpc.service: S3 rpc.system: aws-api Histogram
Value: Sum: 0.3325341 Count: 1 Min: 0.3325341 Max: 0.3325341
(0,5]:1

(2024-07-09T20:51:56.0631967Z, 2024-07-09T20:52:15.4079126Z] rpc.method: GetACL rpc.service: S3 rpc.system: aws-api Histogram
Value: Sum: 0.0383011 Count: 1 Min: 0.0383011 Max: 0.0383011
(0,5]:1

(2024-07-09T20:51:56.0631967Z, 2024-07-09T20:52:15.4079126Z] rpc.method: ListObjects rpc.service: S3 rpc.system: aws-api Histogram
Value: Sum: 0.0329314 Count: 1 Min: 0.0329314 Max: 0.0329314
(0,5]:1

(2024-07-09T20:51:56.0631967Z, 2024-07-09T20:52:15.4079126Z] rpc.method: PutObject rpc.service: S3 rpc.system: aws-api Histogram
Value: Sum: 6.1348382 Count: 35 Min: 0.1319425 Max: 0.4694568
(0,5]:35

(2024-07-09T20:51:56.0631967Z, 2024-07-09T20:52:15.4079126Z] rpc.method: GetObject rpc.service: S3 rpc.system: aws-api Histogram
Value: Sum: 7.1103559 Count: 37 Min: 0.0352235 Max: 3.5720756
(0,5]:37

(2024-07-09T20:51:56.0631967Z, 2024-07-09T20:52:15.4079126Z] rpc.method: ListVersions rpc.service: S3 rpc.system: aws-api Histogram
Value: Sum: 0.0602026 Count: 1 Min: 0.0602026 Max: 0.0602026
(0,5]:1

(2024-07-09T20:51:56.0631967Z, 2024-07-09T20:52:15.4079126Z] rpc.method: DeleteObjects rpc.service: S3 rpc.system: aws-api Histogram
Value: Sum: 0.4776557 Count: 1 Min: 0.4776557 Max: 0.4776557
(0,5]:1

(2024-07-09T20:51:56.0631967Z, 2024-07-09T20:52:15.4079126Z] rpc.method: DeleteBucket rpc.service: S3 rpc.system: aws-api Histogram
Value: Sum: 0.2283794 Count: 1 Min: 0.2283794 Max: 0.2283794
(0,5]:1


Metric Name: client.call.serialization_duration, Unit: s, Meter: AWSSDK.S3
(2024-07-09T20:51:56.0822022Z, 2024-07-09T20:52:15.4079146Z] rpc.method: ListBuckets rpc.service: S3 rpc.system: aws-api Histogram
Value: Sum: 0.0170236 Count: 1 Min: 0.0170236 Max: 0.0170236
(0,5]:1

(2024-07-09T20:51:56.0822022Z, 2024-07-09T20:52:15.4079146Z] rpc.method: PutBucket rpc.service: S3 rpc.system: aws-api Histogram
Value: Sum: 0.0042885 Count: 1 Min: 0.0042885 Max: 0.0042885
(0,5]:1

(2024-07-09T20:51:56.0822022Z, 2024-07-09T20:52:15.4079146Z] rpc.method: GetACL rpc.service: S3 rpc.system: aws-api Histogram
Value: Sum: 0.0008517 Count: 1 Min: 0.0008517 Max: 0.0008517
(0,5]:1

(2024-07-09T20:51:56.0822022Z, 2024-07-09T20:52:15.4079146Z] rpc.method: ListObjects rpc.service: S3 rpc.system: aws-api Histogram
Value: Sum: 0.0009378 Count: 1 Min: 0.0009378 Max: 0.0009378
(0,5]:1

(2024-07-09T20:51:56.0822022Z, 2024-07-09T20:52:15.4079146Z] rpc.method: PutObject rpc.service: S3 rpc.system: aws-api Histogram
Value: Sum: 0.005508099999999999 Count: 35 Min: 2.97E-05 Max: 0.0035104
(0,5]:35

(2024-07-09T20:51:56.0822022Z, 2024-07-09T20:52:15.4079146Z] rpc.method: GetObject rpc.service: S3 rpc.system: aws-api Histogram
Value: Sum: 0.0036232999999999994 Count: 37 Min: 2.89E-05 Max: 0.0017374
(0,5]:37

(2024-07-09T20:51:56.0822022Z, 2024-07-09T20:52:15.4079146Z] rpc.method: ListVersions rpc.service: S3 rpc.system: aws-api Histogram
Value: Sum: 0.0008432 Count: 1 Min: 0.0008432 Max: 0.0008432
(0,5]:1

(2024-07-09T20:51:56.0822022Z, 2024-07-09T20:52:15.4079146Z] rpc.method: DeleteObjects rpc.service: S3 rpc.system: aws-api Histogram
Value: Sum: 0.0021004 Count: 1 Min: 0.0021004 Max: 0.0021004
(0,5]:1

(2024-07-09T20:51:56.0822022Z, 2024-07-09T20:52:15.4079146Z] rpc.method: DeleteBucket rpc.service: S3 rpc.system: aws-api Histogram
Value: Sum: 0.0004062 Count: 1 Min: 0.0004062 Max: 0.0004062
(0,5]:1


Metric Name: client.call.resolve_endpoint_duration, Unit: s, Meter: AWSSDK.S3
(2024-07-09T20:51:56.1469390Z, 2024-07-09T20:52:15.4079161Z] rpc.method: ListBuckets rpc.service: S3 rpc.system: aws-api Histogram
Value: Sum: 0.0671123 Count: 1 Min: 0.0671123 Max: 0.0671123
(0,5]:1

(2024-07-09T20:51:56.1469390Z, 2024-07-09T20:52:15.4079161Z] rpc.method: PutBucket rpc.service: S3 rpc.system: aws-api Histogram
Value: Sum: 0.0012998 Count: 1 Min: 0.0012998 Max: 0.0012998
(0,5]:1

(2024-07-09T20:51:56.1469390Z, 2024-07-09T20:52:15.4079161Z] rpc.method: GetACL rpc.service: S3 rpc.system: aws-api Histogram
Value: Sum: 0.0001225 Count: 1 Min: 0.0001225 Max: 0.0001225
(0,5]:1

(2024-07-09T20:51:56.1469390Z, 2024-07-09T20:52:15.4079161Z] rpc.method: ListObjects rpc.service: S3 rpc.system: aws-api Histogram
Value: Sum: 0.0002505 Count: 1 Min: 0.0002505 Max: 0.0002505
(0,5]:1

(2024-07-09T20:51:56.1469390Z, 2024-07-09T20:52:15.4079161Z] rpc.method: PutObject rpc.service: S3 rpc.system: aws-api Histogram
Value: Sum: 0.0065437 Count: 35 Min: 8.83E-05 Max: 0.00064
(0,5]:35

(2024-07-09T20:51:56.1469390Z, 2024-07-09T20:52:15.4079161Z] rpc.method: GetObject rpc.service: S3 rpc.system: aws-api Histogram
Value: Sum: 0.0061722999999999995 Count: 37 Min: 8.59E-05 Max: 0.0003003
(0,5]:37

(2024-07-09T20:51:56.1469390Z, 2024-07-09T20:52:15.4079161Z] rpc.method: ListVersions rpc.service: S3 rpc.system: aws-api Histogram
Value: Sum: 0.0001288 Count: 1 Min: 0.0001288 Max: 0.0001288
(0,5]:1

(2024-07-09T20:51:56.1469390Z, 2024-07-09T20:52:15.4079161Z] rpc.method: DeleteObjects rpc.service: S3 rpc.system: aws-api Histogram
Value: Sum: 8.92E-05 Count: 1 Min: 8.92E-05 Max: 8.92E-05
(0,5]:1

(2024-07-09T20:51:56.1469390Z, 2024-07-09T20:52:15.4079161Z] rpc.method: DeleteBucket rpc.service: S3 rpc.system: aws-api Histogram
Value: Sum: 8.92E-05 Count: 1 Min: 8.92E-05 Max: 8.92E-05
(0,5]:1


Metric Name: client.call.auth.resolve_identity_duration, Unit: s, Meter: AWSSDK.S3
(2024-07-09T20:51:56.2295096Z, 2024-07-09T20:52:15.4079180Z] rpc.method: ListBuckets rpc.service: S3 rpc.system: aws-api Histogram
Value: Sum: 0.002077 Count: 1 Min: 0.002077 Max: 0.002077
(0,5]:1

(2024-07-09T20:51:56.2295096Z, 2024-07-09T20:52:15.4079180Z] rpc.method: PutBucket rpc.service: S3 rpc.system: aws-api Histogram
Value: Sum: 6.9E-06 Count: 1 Min: 6.9E-06 Max: 6.9E-06
(0,5]:1

(2024-07-09T20:51:56.2295096Z, 2024-07-09T20:52:15.4079180Z] rpc.method: GetACL rpc.service: S3 rpc.system: aws-api Histogram
Value: Sum: 4E-06 Count: 1 Min: 4E-06 Max: 4E-06
(0,5]:1

(2024-07-09T20:51:56.2295096Z, 2024-07-09T20:52:15.4079180Z] rpc.method: ListObjects rpc.service: S3 rpc.system: aws-api Histogram
Value: Sum: 5.5E-06 Count: 1 Min: 5.5E-06 Max: 5.5E-06
(0,5]:1

(2024-07-09T20:51:56.2295096Z, 2024-07-09T20:52:15.4079180Z] rpc.method: PutObject rpc.service: S3 rpc.system: aws-api Histogram
Value: Sum: 0.0002614 Count: 35 Min: 2.6E-06 Max: 2.51E-05
(0,5]:35

(2024-07-09T20:51:56.2295096Z, 2024-07-09T20:52:15.4079180Z] rpc.method: GetObject rpc.service: S3 rpc.system: aws-api Histogram
Value: Sum: 0.00021980000000000006 Count: 37 Min: 2.1E-06 Max: 1.32E-05
(0,5]:37

(2024-07-09T20:51:56.2295096Z, 2024-07-09T20:52:15.4079180Z] rpc.method: ListVersions rpc.service: S3 rpc.system: aws-api Histogram
Value: Sum: 5E-06 Count: 1 Min: 5E-06 Max: 5E-06
(0,5]:1

(2024-07-09T20:51:56.2295096Z, 2024-07-09T20:52:15.4079180Z] rpc.method: DeleteObjects rpc.service: S3 rpc.system: aws-api Histogram
Value: Sum: 4.4E-06 Count: 1 Min: 4.4E-06 Max: 4.4E-06
(0,5]:1

(2024-07-09T20:51:56.2295096Z, 2024-07-09T20:52:15.4079180Z] rpc.method: DeleteBucket rpc.service: S3 rpc.system: aws-api Histogram
Value: Sum: 4E-06 Count: 1 Min: 4E-06 Max: 4E-06
(0,5]:1


Metric Name: client.call.auth.signing_duration, Unit: s, Meter: AWSSDK.S3
(2024-07-09T20:51:56.2412004Z, 2024-07-09T20:52:15.4079196Z] rpc.method: ListBuckets rpc.service: S3 rpc.system: aws-api Histogram
Value: Sum: 0.0859607 Count: 1 Min: 0.0859607 Max: 0.0859607
(0,5]:1

(2024-07-09T20:51:56.2412004Z, 2024-07-09T20:52:15.4079196Z] rpc.method: PutBucket rpc.service: S3 rpc.system: aws-api Histogram
Value: Sum: 0.0006397 Count: 1 Min: 0.0006397 Max: 0.0006397
(0,5]:1

(2024-07-09T20:51:56.2412004Z, 2024-07-09T20:52:15.4079196Z] rpc.method: GetACL rpc.service: S3 rpc.system: aws-api Histogram
Value: Sum: 0.0039421 Count: 1 Min: 0.0039421 Max: 0.0039421
(0,5]:1

(2024-07-09T20:51:56.2412004Z, 2024-07-09T20:52:15.4079196Z] rpc.method: ListObjects rpc.service: S3 rpc.system: aws-api Histogram
Value: Sum: 0.0001217 Count: 1 Min: 0.0001217 Max: 0.0001217
(0,5]:1

(2024-07-09T20:51:56.2412004Z, 2024-07-09T20:52:15.4079196Z] rpc.method: PutObject rpc.service: S3 rpc.system: aws-api Histogram
Value: Sum: 0.0236498 Count: 35 Min: 8.85E-05 Max: 0.0150428
(0,5]:35

(2024-07-09T20:51:56.2412004Z, 2024-07-09T20:52:15.4079196Z] rpc.method: GetObject rpc.service: S3 rpc.system: aws-api Histogram
Value: Sum: 0.006673200000000001 Count: 37 Min: 7.83E-05 Max: 0.0003617
(0,5]:37

(2024-07-09T20:51:56.2412004Z, 2024-07-09T20:52:15.4079196Z] rpc.method: ListVersions rpc.service: S3 rpc.system: aws-api Histogram
Value: Sum: 0.0001037 Count: 1 Min: 0.0001037 Max: 0.0001037
(0,5]:1

(2024-07-09T20:51:56.2412004Z, 2024-07-09T20:52:15.4079196Z] rpc.method: DeleteObjects rpc.service: S3 rpc.system: aws-api Histogram
Value: Sum: 8.81E-05 Count: 1 Min: 8.81E-05 Max: 8.81E-05
(0,5]:1

(2024-07-09T20:51:56.2412004Z, 2024-07-09T20:52:15.4079196Z] rpc.method: DeleteBucket rpc.service: S3 rpc.system: aws-api Histogram
Value: Sum: 8.89E-05 Count: 1 Min: 8.89E-05 Max: 8.89E-05
(0,5]:1


Metric Name: client.call.attempt_duration, Unit: s, Meter: AWSSDK.S3
(2024-07-09T20:51:56.3794607Z, 2024-07-09T20:52:15.4079213Z] rpc.method: ListBuckets rpc.service: S3 rpc.system: aws-api Histogram
Value: Sum: 0.3889079 Count: 1 Min: 0.3889079 Max: 0.3889079
(0,5]:1

(2024-07-09T20:51:56.3794607Z, 2024-07-09T20:52:15.4079213Z] rpc.method: PutBucket rpc.service: S3 rpc.system: aws-api Histogram
Value: Sum: 0.3208213 Count: 1 Min: 0.3208213 Max: 0.3208213
(0,5]:1

(2024-07-09T20:51:56.3794607Z, 2024-07-09T20:52:15.4079213Z] rpc.method: GetACL rpc.service: S3 rpc.system: aws-api Histogram
Value: Sum: 0.0286077 Count: 1 Min: 0.0286077 Max: 0.0286077
(0,5]:1

(2024-07-09T20:51:56.3794607Z, 2024-07-09T20:52:15.4079213Z] rpc.method: ListObjects rpc.service: S3 rpc.system: aws-api Histogram
Value: Sum: 0.0262215 Count: 1 Min: 0.0262215 Max: 0.0262215
(0,5]:1

(2024-07-09T20:51:56.3794607Z, 2024-07-09T20:52:15.4079213Z] rpc.method: PutObject rpc.service: S3 rpc.system: aws-api Histogram
Value: Sum: 6.0725575 Count: 35 Min: 0.1312982 Max: 0.4680492
(0,5]:35

(2024-07-09T20:51:56.3794607Z, 2024-07-09T20:52:15.4079213Z] rpc.method: GetObject rpc.service: S3 rpc.system: aws-api Histogram
Value: Sum: 1.7543711000000002 Count: 37 Min: 0.0337125 Max: 0.1999691
(0,5]:37

(2024-07-09T20:51:56.3794607Z, 2024-07-09T20:52:15.4079213Z] rpc.method: ListVersions rpc.service: S3 rpc.system: aws-api Histogram
Value: Sum: 0.0402997 Count: 1 Min: 0.0402997 Max: 0.0402997
(0,5]:1

(2024-07-09T20:51:56.3794607Z, 2024-07-09T20:52:15.4079213Z] rpc.method: DeleteObjects rpc.service: S3 rpc.system: aws-api Histogram
Value: Sum: 0.4729442 Count: 1 Min: 0.4729442 Max: 0.4729442
(0,5]:1

(2024-07-09T20:51:56.3794607Z, 2024-07-09T20:52:15.4079213Z] rpc.method: DeleteBucket rpc.service: S3 rpc.system: aws-api Histogram
Value: Sum: 0.2266923 Count: 1 Min: 0.2266923 Max: 0.2266923
(0,5]:1


Metric Name: client.call.deserialization_duration, Unit: s, Meter: AWSSDK.S3
(2024-07-09T20:51:56.7941121Z, 2024-07-09T20:52:15.4079231Z] rpc.method: ListBuckets rpc.service: S3 rpc.system: aws-api Histogram
Value: Sum: 0.018165 Count: 1 Min: 0.018165 Max: 0.018165
(0,5]:1

(2024-07-09T20:51:56.7941121Z, 2024-07-09T20:52:15.4079231Z] rpc.method: PutBucket rpc.service: S3 rpc.system: aws-api Histogram
Value: Sum: 0.0001223 Count: 1 Min: 0.0001223 Max: 0.0001223
(0,5]:1

(2024-07-09T20:51:56.7941121Z, 2024-07-09T20:52:15.4079231Z] rpc.method: GetACL rpc.service: S3 rpc.system: aws-api Histogram
Value: Sum: 0.0037079 Count: 1 Min: 0.0037079 Max: 0.0037079
(0,5]:1

(2024-07-09T20:51:56.7941121Z, 2024-07-09T20:52:15.4079231Z] rpc.method: ListObjects rpc.service: S3 rpc.system: aws-api Histogram
Value: Sum: 0.00239 Count: 1 Min: 0.00239 Max: 0.00239
(0,5]:1

(2024-07-09T20:51:56.7941121Z, 2024-07-09T20:52:15.4079231Z] rpc.method: PutObject rpc.service: S3 rpc.system: aws-api Histogram
Value: Sum: 0.0015826000000000002 Count: 35 Min: 8.2E-06 Max: 0.0010906
(0,5]:35

(2024-07-09T20:51:56.7941121Z, 2024-07-09T20:52:15.4079231Z] rpc.method: GetObject rpc.service: S3 rpc.system: aws-api Histogram
Value: Sum: 0.004999700000000001 Count: 35 Min: 3.01E-05 Max: 0.0030071
(0,5]:35

(2024-07-09T20:51:56.7941121Z, 2024-07-09T20:52:15.4079231Z] rpc.method: ListVersions rpc.service: S3 rpc.system: aws-api Histogram
Value: Sum: 0.0178718 Count: 1 Min: 0.0178718 Max: 0.0178718
(0,5]:1

(2024-07-09T20:51:56.7941121Z, 2024-07-09T20:52:15.4079231Z] rpc.method: DeleteObjects rpc.service: S3 rpc.system: aws-api Histogram
Value: Sum: 0.0015635 Count: 1 Min: 0.0015635 Max: 0.0015635
(0,5]:1

(2024-07-09T20:51:56.7941121Z, 2024-07-09T20:52:15.4079231Z] rpc.method: DeleteBucket rpc.service: S3 rpc.system: aws-api Histogram
Value: Sum: 0.0001717 Count: 1 Min: 0.0001717 Max: 0.0001717
(0,5]:1


Metric Name: client.http.bytes_sent, Unit: By, Meter: AWSSDK.S3
(2024-07-09T20:51:57.4209274Z, 2024-07-09T20:52:15.4079237Z] rpc.method: PutObject rpc.service: S3 rpc.system: aws-api server.address: newbucket1902445051.s3.amazonaws.com:443 LongSum
Value: 256496
(2024-07-09T20:51:57.4209274Z, 2024-07-09T20:52:15.4079237Z] rpc.method: DeleteObjects rpc.service: S3 rpc.system: aws-api server.address: newbucket1902445051.s3.amazonaws.com:443 LongSum
Value: 2537

Metric Name: client.http.bytes_received, Unit: By, Meter: AWSSDK.S3
(2024-07-09T20:51:57.4652236Z, 2024-07-09T20:52:15.4079241Z] rpc.method: GetObject rpc.service: S3 rpc.system: aws-api server.address: newbucket1902445051.s3.amazonaws.com:443 LongSum
Value: 250390

Metric Name: client.call.errors, Unit: {error}, Meter: AWSSDK.S3
(2024-07-09T20:52:08.3364408Z, 2024-07-09T20:52:15.4079246Z] exception.type: AmazonS3Exception rpc.method: GetObject rpc.service: S3 rpc.system: aws-api LongSum
Value: 2
```

</details>

<details>
  <summary><b>Examples after exporting to prometheus:</b></summary>

* client.call.duration
image

* client.uptime (uptime is per client, thats why it doesn't show methods)
image

* client.call.attempts
image

* client.call.errors
image

* client.call.attempt_duration
image

* client.call.resolve_endpoint_duration
image

* client.call.serialization_duration
image

* client.call.deserialization_duration
image

* client.call.auth.signing_duration
image

* client.call.auth.resolve_identity_duration
image

* client.http.bytes_sent
image

* client.http.bytes_received
image

</details>

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] My code follows the code style of this project
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly
- [x] I have read the **README** document
- [ ] I have added tests to cover my changes
- [ ] All new and existing tests passed

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

[issues]: https://github.com/aws/aws-sdk-net/issues
[license]: http://aws.amazon.com/apache2.0/
[cla]: http://en.wikipedia.org/wiki/Contributor_License_Agreement

### CODE REVIEW

File: sdk/src/Core/Amazon.Runtime/AmazonServiceClient.cs [hunk 1]
```diff
@@ -411,6 +416,9 @@ protected virtual void Dispose(bool disposing)
                 if (RuntimePipeline != null)
                     RuntimePipeline.Dispose();
 
+                if(_uptimeMetricMeasurer != null)
+                    _uptimeMetricMeasurer.Dispose();
+
```
Comment (boblodgett) [inline sdk/src/Core/Amazon.Runtime/AmazonServiceClient.cs:421]: Should set the variable to NULL after dispose: https://learn.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-dispose#the-disposebool-method-overload . Also a simplified way is:

```
                _safeHandle?.Dispose();
                _safeHandle = null;

```

File: sdk/src/Core/Amazon.Runtime/Pipeline/Handlers/Marshaller.cs [hunk 2]
```diff
@@ -77,28 +79,31 @@ public override IAsyncResult InvokeAsync(IAsyncExecutionContext executionContext
...
+            {
+                var requestContext = executionContext.RequestContext;
+                requestContext.Request = requestContext.Marshaller.Marshall(requestContext.OriginalRequest);
+                requestContext.Request.AuthenticationRegion = requestContext.ClientConfig.AuthenticationRegion;
 
 #if NETSTANDARD
-            var method = requestContext.Request.HttpMethod.ToUpperInvariant();
+                var method = requestContext.Request.HttpMethod.ToUpperInvariant();
 #else
             var method = requestContext.Request.HttpMethod.ToUpper(CultureInfo.InvariantCulture);
```
Comment (boblodgett) [inline sdk/src/Core/Amazon.Runtime/Pipeline/Handlers/Marshaller.cs:91]: If you are going to correct indentation in the prior case it should be done here as well.

File: sdk/src/Core/Amazon.Runtime/Pipeline/Handlers/MetricsHandler.cs [hunk 3]
```diff
@@ -40,7 +41,8 @@ public override void InvokeSync(IExecutionContext executionContext)
             var operationName = AWSSDKUtils.ExtractOperationName(executionContext.RequestContext.RequestName);
             var spanName = $"{executionContext.RequestContext.ServiceMetaData.ServiceId}.{operationName}";
             var span = TracingUtilities.CreateSpan(executionContext.RequestContext, spanName);
-
+            var callDurationMetricsMeasurer = MetricsUtilities.MeasureDuration(executionContext.RequestContext, TelemetryConstants.CallDurationMetricName);
```
Comment (boblodgett) [inline sdk/src/Core/Amazon.Runtime/Pipeline/Handlers/MetricsHandler.cs:44]: Is there a case where MeasureDuration can throw an exception outside of the try? If so the span.dispose won't be called in the finally and you already created the span.

File: sdk/src/Core/Amazon.Runtime/Pipeline/HttpHandler/HttpHandler.cs [hunk 4]
```diff
@@ -136,15 +138,35 @@ private static void CompleteFailedRequest(IHttpRequest<TRequestContent> httpRequ
...
+
+            var metricsAttributes = new Attributes();
+            metricsAttributes.Set(TelemetryConstants.ServerAddressAttributeKey, $"{request.Endpoint.Host}:{request.Endpoint.Port.ToString()}");
+
+            traceSpan.SetAttribute(TelemetryConstants.HTTPMethodAttributeKey, request.HttpMethod);
             traceSpan.SetAttribute(TelemetryConstants.HTTPStatusCodeAttributeKey, ((int)response.StatusCode));
 
             var contentLengthHeader = request.GetHeaderValue(HeaderKeys.ContentLengthHeader);
-            if (!string.IsNullOrEmpty(contentLengthHeader) && long.TryParse(contentLengthHeader, out var contentLength))
+            if (long.TryParse(contentLengthHeader, out var contentLength) && contentLength > 0)
```
Comment (boblodgett) [inline sdk/src/Core/Amazon.Runtime/Pipeline/HttpHandler/HttpHandler.cs:152]: What is the reason of treating a missing content-length the same as one that is zero? Is this correct in all cases or what case supports this change?

File: sdk/src/Core/Amazon.Runtime/Pipeline/RetryHandler/RetryHandler.cs [hunk 5]
```diff
@@ -90,6 +92,8 @@ public override void InvokeSync(IExecutionContext executionContext)
                     {
                         requestContext.Retries++;
                         requestContext.Metrics.SetCounter(Metric.AttemptCount, requestContext.Retries);
+                        MetricsUtilities.AddMonotonicCounterValue(requestContext, TelemetryConstants.CallAttemptsMetricName, TelemetryConstants.AttemptUnitName, 1);
```
Comment (boblodgett) [inline sdk/src/Core/Amazon.Runtime/Pipeline/RetryHandler/RetryHandler.cs:95]: Why `1` or what does it represent? I see this in a few places seemingly as a default value.

File: sdk/src/Core/Amazon.Runtime/Telemetry/Metrics/MetricsUtilities.cs [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+
+        /// <summary>
+        /// Records a value using a MonotonicCounter.
+        /// </summary>
+        /// <param name="requestContext">The request context object.</param>
+        /// <param name="metricName">The name of the metric to record.</param>
+        /// <param name="unit">The unit of the metric.</param>
+        /// <param name="value">The value to be recorded by the MonotonicCounter.</param>
+        /// <param name="initialAttributes">The attributes associated with the metric.</param>
+        public static void AddMonotonicCounterValue(IRequestContext requestContext, string metricName, string unit, long value, Attributes initialAttributes = null)
```
Comment (boblodgett) [inline sdk/src/Core/Amazon.Runtime/Telemetry/Metrics/MetricsUtilities.cs:47]: Is `1` the default for `value`?

File: sdk/src/Core/Amazon.Runtime/Telemetry/TelemetryProvider.cs [hunk 7]
```diff
@@ -27,25 +28,41 @@ public abstract class TelemetryProvider
...
         /// <summary>
         /// Gets the <see cref="TracerProvider"/> used to create new tracers.
         /// </summary>
-        public abstract TracerProvider TracerProvider { get; protected set; }
+        public TracerProvider TracerProvider { get; private set; }
+
+        public TelemetryProvider(MeterProvider meterProvider, TracerProvider tracerProvider)
+        {
+            MeterProvider = meterProvider;
+            TracerProvider = tracerProvider;
```
Comment (boblodgett) [inline sdk/src/Core/Amazon.Runtime/Telemetry/TelemetryProvider.cs:41]: Should this be calling the Register* methods instead so it does the null check or is it valid to be able to specify null here but not if the Register* methods are called?

**Review by boblodgett — APPROVED** [review]

**Review by afroz429 — APPROVED** [review]

**Pull request merged**